### PR TITLE
AiiDAlab-ELN ingegration.

### PR DESCRIPTION
**Pull request by yakutovicha**

### CODE REVIEW

File: aiidalab_widgets_base/elns.py [hunk 1]
```diff
@@ -0,0 +1,361 @@
...
+            self.tab.set_title(1, "Geometry analysis")
+            self.tab.set_title(2, "Isotherm")
+
+            self.tab_selection.options = [
+                (
+                    "Geometry Optimization - typically this is the first step needed to find optimal positions of atoms in the unit cell.",
+                    0,
+                ),
+                (
+                    "Geometry analysis - calculate parameters describing the geometry of a porous material.",
```
Comment (kjappelbaum) [inline aiidalab_widgets_base/elns.py:109]:
```suggestion
                    "Pore analysis - calculate parameters describing the pore geometry of a porous material.",
```

File: aiidalab_widgets_base/elns.py [hunk 2]
```diff
@@ -0,0 +1,361 @@
...
+            icon="https://gitlab.com/QEF/q-e/raw/develop/logo.jpg",
+            link=f"https://aiidalab-demo.materialscloud.org/user-redirect/apps/apps/quantum-espresso/qe.ipynb?structure_uuid={self.node.uuid}",
+            description="Optimize atomic positions and/or unit cell employing Quantum ESPRESSO. Quantum ESPRESSO is preferable for small structures with no cell dimensions larger than 15 Å. Additionally, you can choose to compute electronic properties of the material such as band structure and density of states.",
+        ).to_html_string()
+
+        geo_opt.value += AppIcon(
+            icon="https://raw.githubusercontent.com/lsmo-epfl/aiidalab-epfl-lsmo/develop/miscellaneous/logos/LSMO.png",
+            link=f"https://aiidalab-demo.materialscloud.org/user-redirect/apps/apps/aiidalab-lsmo/multistage_geo_opt_ddec.ipynb?structure_uuid={self.node.uuid}",
+            description="Optimize atomic positions and unit cell with CP2K. CP2K is very efficient for large and/or porous structures. A structure is considered large when any cell dimension is larger than 15 Å. Additionally, you can choose to assign point charges to the atoms using DDEC.",
```
Comment (kjappelbaum) [inline aiidalab_widgets_base/elns.py:133]:
```suggestion
            description="Optimize atomic positions and unit cell with CP2K. CP2K is very efficient for large (any cell dimension is larger than 15 Å) and/or porous structures. Additionally, you can choose to assign point charges to the atoms using DDEC.",
```

File: aiidalab_widgets_base/elns.py [hunk 3]
```diff
@@ -0,0 +1,361 @@
...
+        ).to_html_string()
+
+        return geo_opt
+
+    def get_isotherm_tab(self):
+        isotherm = ipw.HTML()
+        isotherm.value += AppIcon(
+            icon="https://raw.githubusercontent.com/lsmo-epfl/aiidalab-epfl-lsmo/develop/miscellaneous/logos/LSMO.png",
+            link=f"https://aiidalab-demo.materialscloud.org/user-redirect/apps/apps/aiidalab-lsmo/compute_isotherm.ipynb?structure_uuid={self.node.uuid}",
+            description="Compute adsorption isotherm of the selected material using the RASPA code. Typically, one needs to optimize geometry and compute the charges of material before isotherm. However, if this is already done, you can go for it.",
```
Comment (kjappelbaum) [inline aiidalab_widgets_base/elns.py:143]:
```suggestion
            description="Compute adsorption isotherm of the selected material using the RASPA code. Typically, one needs to optimize geometry and compute the charges of material before computing the isotherm. However, if this is already done, you can go for it.",
```

File: aiidalab_widgets_base/elns.py [hunk 4]
```diff
@@ -0,0 +1,361 @@
...
+        ).to_html_string()
+
+        return isotherm
+
+    def get_geometry_analysis_tab(self):
+        geo_analysis = ipw.HTML()
+        geo_analysis.value += AppIcon(
+            icon="https://raw.githubusercontent.com/lsmo-epfl/aiidalab-epfl-lsmo/develop/miscellaneous/logos/LSMO.png",
+            link=f"https://aiidalab-demo.materialscloud.org/user-redirect/apps/apps/aiidalab-lsmo/pore_analysis.ipynb?structure_uuid={self.node.uuid}",
+            description="Perform geometry analysis of a material employing Zeo++ code.",
```
Comment (kjappelbaum) [inline aiidalab_widgets_base/elns.py:153]:
```suggestion
            description="Calculate descriptors for the pore geometry using the Zeo++ code.",
```

File: aiidalab_widgets_base/elns.py [hunk 5]
```diff
@@ -0,0 +1,361 @@
...
+        self._output = ipw.Output()
+
+        # Communicate to the user if something isn't right.
+        error_message = ipw.HTML()
+        super().__init__(children=[error_message], **kwargs)
+
+        eln = connect_to_eln(**kwargs)
+
+        if eln is None:
+            error_message.value = f"""Warning! The access to ELN {kwargs['eln_name']} is not configured. Please follow <a href="https://aiidalab-demo.materialscloud.org/user/aliaksandr.yakutovich@epfl.ch/apps/apps/aiidalab-widgets-base/eln_configure.ipynb" target="_blank">the link</a> to configure it."""
```
Comment (kjappelbaum) [inline aiidalab_widgets_base/elns.py:173]: we should not hardcode the username, but i cannot make a suggestion as I do not know how to inspect what the username is :D

File: aiidalab_widgets_base/elns.py [hunk 6]
```diff
@@ -0,0 +1,361 @@
...
+                info = {}
+
+        self.eln.set_sample_config(**info)
+
+    def send_to_eln(self, _=None):
+        if self.eln.is_connected:
+            self.error_message.value = ""
+            self.eln.export_data_object(data_object=self.node)
+        else:
+            self.error_message.value = """Warning! The access to ELN is not configured. Please follow <a href="https://aiidalab-demo.materialscloud.org/user/aliaksandr.yakutovich@epfl.ch/apps/apps/aiidalab-widgets-base/eln_configure.ipynb" target="_blank">the link</a> to configure it."""
```
Comment (kjappelbaum) [inline aiidalab_widgets_base/elns.py:239]: same as the other comment about the username

File: aiidalab_widgets_base/elns.py [hunk 7]
```diff
@@ -0,0 +1,361 @@
...
+
+class ElnConfigureWidget(ipw.VBox):
+    selected = traitlets.Dict()
+
+    def __init__(self, **kwargs):
+        self._output = ipw.Output()
+
+        self.eln_instance = ipw.Dropdown(
+            description="ELN:",
+            options=("Setup new ELN", {}),
```
Comment (kjappelbaum) [inline aiidalab_widgets_base/elns.py:256]:
```suggestion
            options=("Set up new ELN", {}),
```

File: aiidalab_widgets_base/elns.py [hunk 8]
```diff
@@ -0,0 +1,361 @@
...
+    @traitlets.observe("node")
+    def _observe_node(self, change):
+        if change["new"]:
+            self.tab.children = [
+                self.get_geo_opt_tab(),
+                self.get_geometry_analysis_tab(),
+                self.get_isotherm_tab(),
+            ]
+            self.tab.set_title(0, "Geometry Optimization")
+            self.tab.set_title(1, "Geometry analysis")
```
Comment (kjappelbaum) [inline aiidalab_widgets_base/elns.py:100]:
```suggestion
            self.tab.set_title(1, "Pore Geometry Analysis")
```

File: aiidalab_widgets_base/elns.py [hunk 9]
```diff
@@ -0,0 +1,361 @@
...
+            self.tab.set_title(2, "Isotherm")
```
Comment (kjappelbaum) [inline aiidalab_widgets_base/elns.py:101]:
```suggestion
            self.tab.set_title(2, "Gas Adsorption Isotherm")
```

File: aiidalab_widgets_base/elns.py [hunk 10]
```diff
@@ -0,0 +1,361 @@
...
+        elns.pop(self.eln_instance.label, None)
+        if "default" in elns and elns["default"] not in elns:
+            elns.pop("default")
+
+        with open(ELN_CONFIG, "w") as file:
+            json.dump(elns, file, indent=4)
+        self.update_list_of_elns()
+
+    def check_connection(self, _=None):
+        print("Not implemented :(")
```
Comment (kjappelbaum) [inline aiidalab_widgets_base/elns.py:337]: Should this rather raise `NotImplementedError`?

**Comment by kjappelbaum** [comment]
it seems to me that `ipytree` needs to be added to the dependencies

**Comment by kjappelbaum** [comment]
and we need to make a release of `aiidalab_eln` and then also add this to the dependencies (shall i do this, maybe first approve the PR that adds the workflow for this)

File: aiidalab_widgets_base/elns.py [hunk 11]
```diff
@@ -0,0 +1,361 @@
...
+from aiida.orm import Node, QueryBuilder
+from IPython.display import clear_output, display
+from aiidalab_eln import get_eln_connector
+
+ELN_CONFIG = Path.home() / ".aiidalab" / "aiidalab-eln-config.json"
+
+
+def connect_to_eln(eln_name=None, **kwargs):
+    try:
+        with open(ELN_CONFIG, "r") as file:
```
Comment (kjappelbaum) [inline aiidalab_widgets_base/elns.py:14]: this fails, for example, if the `.aiidalab` folder does not exist.

File: aiidalab_widgets_base/elns.py (hunk 7 above)
Comment (kjappelbaum) [reply aiidalab_widgets_base/elns.py:256]: as this is currently an option, clicking on "set as default" will set this as default. 
What I would maybe add is a check that this key is in the dict with url and token before it is set as default

File: aiidalab_widgets_base/elns.py [hunk 12]
```diff
@@ -0,0 +1,361 @@
...
+                (
+                    "Isotherm - compute adsorption isotherm of a small molecules in the selected material. ",
+                    2,
+                ),
+            ]
+            ipw.link((self.tab, "selected_index"), (self.tab_selection, "value"))
+        else:
+            self.tab.children = []
+
+    def get_geo_opt_tab(self):
```
Comment (kjappelbaum) [inline aiidalab_widgets_base/elns.py:121]: shall the notebooks take the `UUID` of the `StructureData` as url parameter and then users won't have to do this manual step and have the structure directly in the optimization app? Same probably applies for the other notebooks.

Comment (yakutovicha) [reply aiidalab_widgets_base/elns.py]: this is already the case.

File: aiidalab_widgets_base/elns.py (hunk 10 above)
Comment (yakutovicha) [reply aiidalab_widgets_base/elns.py:370]: I've implemented a proper check.

File: aiidalab_widgets_base/elns.py (hunk 11 above)
Comment (yakutovicha) [reply aiidalab_widgets_base/elns.py:68]: addressed in 041f5c822bc2612bf3b08a941e779825cb3442e1

File: aiidalab_widgets_base/elns.py (hunk 5 above)
Comment (yakutovicha) [reply aiidalab_widgets_base/elns.py:206]: fixed.

File: aiidalab_widgets_base/elns.py (hunk 6 above)
Comment (yakutovicha) [reply aiidalab_widgets_base/elns.py:272]: fixed.

**Comment by yakutovicha** [comment]
closing this in favour of #213, #218, #220.

**Pull request closed without merging**